File: android/app/src/main/kotlin/com/stillmoment/domain/services/TimerReducer.kt
package com.stillmoment.domain.services

import com.stillmoment.domain.models.MeditationSettings
import com.stillmoment.domain.models.TimerAction
import com.stillmoment.domain.models.TimerEffect
import com.stillmoment.domain.models.TimerState

/**
 * Pure effect mapper for timer actions.
 *
 * Maps actions to effects based on the current timer state and settings.
 * Contains no side effects and no state mutations - all I/O is represented as effects.
 * State is managed directly by the ViewModel via MeditationTimer.
 */
object TimerReducer {
    /**
     * Maps an action to effects based on current timer state and settings.
     *
     * @param action Action to process
     * @param timerState Current timer state (.Idle when no timer exists)
     * @param selectedMinutes Currently selected duration in minutes
     * @param settings Current meditation settings (for effect parameters)
     * @return Effects to execute
     */
    fun reduce(
        action: TimerAction,
        timerState: TimerState,
        selectedMinutes: Int,
        settings: MeditationSettings,
    ): List<TimerEffect> {
        return when (action) {
            is TimerAction.StartPressed -> reduceStartPressed(selectedMinutes, settings)
            is TimerAction.ResetPressed -> reduceResetPressed(timerState)
            is TimerAction.PreparationFinished -> reducePreparationFinished(settings)
            is TimerAction.StartGongFinished -> reduceStartGongFinished(timerState, settings)
            is TimerAction.TimerCompleted -> reduceTimerCompleted(settings)
            is TimerAction.EndGongFinished -> reduceEndGongFinished(timerState)
            is TimerAction.IntervalGongTriggered -> reduceIntervalGongTriggered(settings)
        }
    }

    // MARK: - Control Actions

    private fun reduceStartPressed(selectedMinutes: Int, settings: MeditationSettings): List<TimerEffect> {
        if (selectedMinutes <= 0) {
            return emptyList()
        }

        // Determine preparation time: skip if disabled (0), otherwise use configured value
        val preparationTime = if (settings.preparationTimeEnabled) {
            settings.preparationTimeSeconds
        } else {
            0
        }

        val updatedSettings = settings.copy(durationMinutes = selectedMinutes)

        // Background audio never starts here. It starts when the start gong finishes
        // (in reduceStartGongFinished). Always use "silent" for foreground service start —
        // background audio is updated later.
        return listOf(
            TimerEffect.StartForegroundService(
                "silent",
                settings.backgroundSoundVolume,
                settings.gongSoundId,
                settings.gongVolume
            ),
            TimerEffect.StartTimer(selectedMinutes, preparationTime),
            TimerEffect.SaveSettings(updatedSettings)
        )
    }

    private fun reduceResetPressed(timerState: TimerState): List<TimerEffect> {
        if (timerState == TimerState.Idle) {
            return emptyList()
        }

        return listOf(
            TimerEffect.StopForegroundService,
            TimerEffect.ResetTimer
        )
    }

    // MARK: - Timer Update Actions

    private fun reducePreparationFinished(settings: MeditationSettings): List<TimerEffect> {
        // Play start gong. Background audio decision is deferred to startGongFinished.
        return listOf(TimerEffect.PlayStartGong(settings.gongSoundId, settings.gongVolume))
    }

    private fun reduceStartGongFinished(timerState: TimerState, settings: MeditationSettings): List<TimerEffect> {
        if (timerState != TimerState.StartGong) {
            return emptyList()
        }

        return listOf(
            TimerEffect.TransitionToRunning,
            TimerEffect.StartBackgroundAudio(settings.backgroundSoundId, settings.backgroundSoundVolume)
        )
    }

    private fun reduceTimerCompleted(settings: MeditationSettings): List<TimerEffect> {
        return listOf(
            TimerEffect.PlayCompletionSound(settings.gongSoundId, settings.gongVolume)
        )
    }

    private fun reduceEndGongFinished(timerState: TimerState): List<TimerEffect> {
        if (timerState != TimerState.EndGong) {
            return emptyList()
        }

        return listOf(TimerEffect.TransitionToCompleted, TimerEffect.StopForegroundService)
    }

    // MARK: - Interval Gong Actions

    private fun reduceIntervalGongTriggered(settings: MeditationSettings): List<TimerEffect> {
        if (!settings.intervalGongsEnabled) {
            return emptyList()
        }
        return listOf(
            TimerEffect.PlayIntervalGong(settings.intervalSoundId, settings.intervalGongVolume)
        )
    }
}

File: ios/StillMoment/Domain/Services/TimerReducer.swift
//
//  TimerReducer.swift
//  Still Moment
//
//  Domain Service - Pure Effect Mapper for Timer Actions
//

import Foundation

/// Pure effect mapper for timer actions
///
/// Maps actions to effects based on the current timer state and settings.
/// Contains no side effects and no state mutations — all I/O is represented as effects.
/// State is managed directly by the ViewModel via MeditationTimer.
enum TimerReducer {
    /// Maps an action to effects based on current timer state and settings
    ///
    /// - Parameters:
    ///   - action: Action to process
    ///   - timerState: Current timer state (.idle when no timer exists)
    ///   - selectedMinutes: Currently selected duration in minutes
    ///   - settings: Current meditation settings (for effect parameters)
    /// - Returns: Effects to execute
    static func reduce(
        action: TimerAction,
        timerState: TimerState,
        selectedMinutes: Int,
        settings: MeditationSettings
    ) -> [TimerEffect] {
        switch action {
        case .startPressed:
            self.reduceStartPressed(selectedMinutes: selectedMinutes)
        case .resetPressed:
            self.reduceResetPressed(timerState: timerState)
        case .preparationFinished:
            self.reducePreparationFinished()
        case .startGongFinished:
            self.reduceStartGongFinished(timerState: timerState, settings: settings)
        case .timerCompleted:
            self.reduceTimerCompleted()
        case .endGongFinished:
            self.reduceEndGongFinished(timerState: timerState)
        case .intervalGongTriggered:
            self.reduceIntervalGong(settings: settings)
        }
    }

    // MARK: - Control Actions

    private static func reduceStartPressed(selectedMinutes: Int) -> [TimerEffect] {
        guard selectedMinutes > 0 else {
            return []
        }

        // Background audio never starts here. It starts when the start gong finishes.
        return [
            .activateTimerSession,
            .startTimer(durationMinutes: selectedMinutes)
        ]
    }

    private static func reduceResetPressed(timerState: TimerState) -> [TimerEffect] {
        guard timerState != .idle else {
            return []
        }
        return [.stopBackgroundAudio, .resetTimer, .clearTimer, .deactivateTimerSession]
    }

    // MARK: - Timer Update Actions

    private static func reducePreparationFinished() -> [TimerEffect] {
        // Play start gong. Background audio decision is deferred to startGongFinished.
        [.playStartGong]
    }

    private static func reduceStartGongFinished(
        timerState: TimerState,
        settings: MeditationSettings
    ) -> [TimerEffect] {
        guard timerState == .startGong else {
            return []
        }

        return [
            .beginRunningPhase,
            .startBackgroundAudio(
                soundId: settings.backgroundSoundId,
                volume: settings.backgroundSoundVolume
            )
        ]
    }

    private static func reduceTimerCompleted() -> [TimerEffect] {
        // Keep-alive stays active during endGong — deactivation happens in reduceEndGongFinished
        [.playCompletionSound, .stopBackgroundAudio]
    }

    private static func reduceEndGongFinished(timerState: TimerState) -> [TimerEffect] {
        guard timerState == .endGong else {
            return []
        }

        return [.transitionToCompleted, .deactivateTimerSession]
    }

    // MARK: - Interval Gong Actions

    /// Handles interval gong triggered by tick() domain event.
    /// Deduplication is handled by MeditationTimer.tick() via lastIntervalGongAt.
    private static func reduceIntervalGong(settings: MeditationSettings) -> [TimerEffect] {
        guard settings.intervalGongsEnabled else {
            return []
        }
        return [
            .playIntervalGong(
                soundId: settings.intervalSoundId,
                volume: settings.intervalGongVolume
            )
        ]
    }
}
